lib/inc/puppetdb-cli/puppetdb-cli.hpp

```diff
@@ -28,10 +28,30 @@ class LIBPUPPETDB_EXPORT PuppetDBConn  {
     PuppetDBConn();
 
     /**
+     * Construct a PuppetDB connection using only config flags
+     * @param urls string of the urls of PuppetDB.
+     * @param cacert string cacert to use for curl.
+     * @param cert client cert to use for curl.
+     * @param key client private key to use for curl.
+     */
+    PuppetDBConn(const std::string& urls,
```

Comment:
I don't think using structs is abnormal however, quite normal in a C/C++ world I would argue ... the problem you avoid is the extrapolation of extra parameters as this gets more and more complex over time. At five parameters, this is getting quite long also, and it raised alarm bells in my head. With a struct you can add fields, deprecate fields and all kinds of magic like that without affecting the API, with positional arguments this becomes harder and you can box yourself in easier.

Also, be mindful that these parameters are wrapped and passed to other functions like you say, you can pass each one individually right, or you can just pass the struct pointer.

Most of my concern is that I can see this thing becoming the pattern for an external api library if we keep going, and exposing this to other consumers concerns me as it stands.